StarryExpanse/RivenGameState.cpp

```diff
@@ -1,16 +1,53 @@
 #include "RivenGameState.h"
+#include "SaveGame/RivenSaveGame.h"
+#include "Interfaces/RivenSaveGameAware.h"
 #include "Engine/Engine.h"
 #include "Kismet/GameplayStatics.h"
 #include "RivenGameInstance.h"
-#include "SaveGame/RivenSaveGame.h"
 #include "StarryExpanse.h"
+#include "Runtime/Core/Public/UObject/WeakObjectPtrTemplates.h"
+#include "Runtime/Core/Public/GenericPlatform/GenericPlatform.h"
 #include "UObject/ConstructorHelpers.h"
 
 ARivenGameState::ARivenGameState() : Super() {}
 
 void ARivenGameState::OnConstruction(const FTransform &Transform) {
   auto initialSavegame = NewObject<URivenSaveGame>();
+  initialSavegame->SetSubscriber(this);
   this->Instantaneous_SaveGame = initialSavegame;
   auto gameInstance = GetWorld()->GetGameInstance<URivenGameInstance>();
   gameInstance->Last_Savable_SaveGame = initialSavegame;
 }
+
+void ARivenGameState::SubscribeActorToSavegame(AActor* actor) {
+  if (actor != nullptr) {
+    if (actor->GetClass()->ImplementsInterface(URivenSavegameAware::StaticClass())) {
+      uint32 uuid = actor->GetUniqueID();
+      SubscribedToSavegame.Add(uuid, TWeakObjectPtr<AActor>(actor));
+    }
+  }
+}
+
+void ARivenGameState::UnsubscribeActorFromSavegame(AActor* actor) {
+  if (actor != nullptr) {
+    uint32 uuid = actor->GetUniqueID();
+    SubscribedToSavegame.FindAndRemoveChecked(uuid);
+  }
+}
+
+void ARivenGameState::NotifySubscribersOfChange(URivenSaveGame *OldSaveGame) {
+  for (auto& Elem : SubscribedToSavegame) {
+    TWeakObjectPtr<AActor> subscriberPointer = Elem.Value;
+    if (
+      subscriberPointer.IsValid() &&
```

Comment:
I would move the IsValid() check up into SubscribeActorToSavegame() and turn both of these into `check`'s.

Reply:
Oh - nevermind. It's a weak ptr which can become NULL.

Reply:
Yep!